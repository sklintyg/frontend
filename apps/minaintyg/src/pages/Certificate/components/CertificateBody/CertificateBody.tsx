import { IDSExpandable } from '@frontend/ids-react-ts'
import parse, { DOMNode, Element, domToReact } from 'html-react-parser'
import { createElement } from 'react'
import { CertificateContent } from '../../../../schema/certificate.schema'

const options = {
  replace: (domNode: DOMNode) => {
    if (domNode instanceof Element) {
      const { name, children } = domNode
      const match = name.match(/^h(\d)$/)
      if (match) {
        return createElement(match[0], { className: `ids-heading-${match[1]}` }, domToReact(children))
      }
      if (name === 'table') {
        return <table className="ids-table">{domToReact(children)}</table>
      }
    }
    return undefined
  },
}

export function CertificateBody({ content }: { content: CertificateContent[] }) {
  return (
    <>
      {content.map(({ heading, body }) => {
        const parsedBody = parse(body, options)
        return (
          <section key={heading} className="md:mb-10">
            <IDSExpandable className="md:hidden" headline={heading}>
              {parsedBody}
            </IDSExpandable>
            <div className="hidden md:block">
              <h2 className="ids-heading-2">{heading}</h2>
              {parsedBody}
            </div>
          </section>
        )
      })}
    </>
  )
}
