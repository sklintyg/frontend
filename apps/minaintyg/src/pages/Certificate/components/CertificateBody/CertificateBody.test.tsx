import { render } from '@testing-library/react'
import { CertificateBody } from './CertificateBody'

it('Should render html as expected', () => {
  const content = [{ heading: 'Rubrik 1', body: '<p>Test</p>' }]
  const { container } = render(<CertificateBody content={content} />)
  expect(container).toMatchSnapshot()
})

it('Should render table as expected', () => {
  const content = [{ heading: 'Rubrik 1', body: '<table><tbody><tr><td>Test</td></tr></tbody></table>' }]
  const { container } = render(<CertificateBody content={content} />)
  expect(container).toMatchSnapshot()
})

it.each(Array.from({ length: 6 }, (_, index) => `h${index + 1}`))('Should render %s heading as expected', (tag) => {
  const content = [{ heading: 'Rubrik 1', body: `<${tag}>heading</${tag}>` }]
  const { container } = render(<CertificateBody content={content} />)
  expect(container).toMatchSnapshot()
})
