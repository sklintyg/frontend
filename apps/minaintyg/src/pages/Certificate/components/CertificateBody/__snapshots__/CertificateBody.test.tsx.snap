// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render h1 heading as expected 1`] = `
<div>
  <div
    class="border-b border-neutral-90"
  >
    <div
      class="ids-accordion"
    >
      <div
        aria-controls=":r2:"
        aria-expanded="false"
        class="ids-accordion__button ids-accordion__button--m"
        role="button"
        tabindex="0"
      >
        <div
          class="ids-accordion__headline ids-accordion__headline-1 ids-accordion__headline--m"
        >
          <h2>
            Rubrik 1
          </h2>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="ids-accordion__content"
        id=":r2:"
      >
        <h1
          class="ids-heading-l"
        >
          heading
        </h1>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render h2 heading as expected 1`] = `
<div>
  <div
    class="border-b border-neutral-90"
  >
    <div
      class="ids-accordion"
    >
      <div
        aria-controls=":r3:"
        aria-expanded="false"
        class="ids-accordion__button ids-accordion__button--m"
        role="button"
        tabindex="0"
      >
        <div
          class="ids-accordion__headline ids-accordion__headline-1 ids-accordion__headline--m"
        >
          <h2>
            Rubrik 1
          </h2>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="ids-accordion__content"
        id=":r3:"
      >
        <h2
          class="ids-heading-m"
        >
          heading
        </h2>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render h3 heading as expected 1`] = `
<div>
  <div
    class="border-b border-neutral-90"
  >
    <div
      class="ids-accordion"
    >
      <div
        aria-controls=":r4:"
        aria-expanded="false"
        class="ids-accordion__button ids-accordion__button--m"
        role="button"
        tabindex="0"
      >
        <div
          class="ids-accordion__headline ids-accordion__headline-1 ids-accordion__headline--m"
        >
          <h2>
            Rubrik 1
          </h2>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="ids-accordion__content"
        id=":r4:"
      >
        <h3
          class="ids-heading-s"
        >
          heading
        </h3>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render h4 heading as expected 1`] = `
<div>
  <div
    class="border-b border-neutral-90"
  >
    <div
      class="ids-accordion"
    >
      <div
        aria-controls=":r5:"
        aria-expanded="false"
        class="ids-accordion__button ids-accordion__button--m"
        role="button"
        tabindex="0"
      >
        <div
          class="ids-accordion__headline ids-accordion__headline-1 ids-accordion__headline--m"
        >
          <h2>
            Rubrik 1
          </h2>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="ids-accordion__content"
        id=":r5:"
      >
        <h4
          class="ids-heading-xs"
        >
          heading
        </h4>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render h5 heading as expected 1`] = `
<div>
  <div
    class="border-b border-neutral-90"
  >
    <div
      class="ids-accordion"
    >
      <div
        aria-controls=":r6:"
        aria-expanded="false"
        class="ids-accordion__button ids-accordion__button--m"
        role="button"
        tabindex="0"
      >
        <div
          class="ids-accordion__headline ids-accordion__headline-1 ids-accordion__headline--m"
        >
          <h2>
            Rubrik 1
          </h2>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="ids-accordion__content"
        id=":r6:"
      >
        <h5
          class="ids-heading-xs"
        >
          heading
        </h5>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render h6 heading as expected 1`] = `
<div>
  <div
    class="border-b border-neutral-90"
  >
    <div
      class="ids-accordion"
    >
      <div
        aria-controls=":r7:"
        aria-expanded="false"
        class="ids-accordion__button ids-accordion__button--m"
        role="button"
        tabindex="0"
      >
        <div
          class="ids-accordion__headline ids-accordion__headline-1 ids-accordion__headline--m"
        >
          <h2>
            Rubrik 1
          </h2>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="ids-accordion__content"
        id=":r7:"
      >
        <h6
          class="ids-heading-xs"
        >
          heading
        </h6>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render html as expected 1`] = `
<div>
  <div
    class="border-b border-neutral-90"
  >
    <div
      class="ids-accordion"
    >
      <div
        aria-controls=":r0:"
        aria-expanded="false"
        class="ids-accordion__button ids-accordion__button--m"
        role="button"
        tabindex="0"
      >
        <div
          class="ids-accordion__headline ids-accordion__headline-1 ids-accordion__headline--m"
        >
          <h2>
            Rubrik 1
          </h2>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="ids-accordion__content"
        id=":r0:"
      >
        <p>
          Test
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render table as expected 1`] = `
<div>
  <div
    class="border-b border-neutral-90"
  >
    <div
      class="ids-accordion"
    >
      <div
        aria-controls=":r1:"
        aria-expanded="false"
        class="ids-accordion__button ids-accordion__button--m"
        role="button"
        tabindex="0"
      >
        <div
          class="ids-accordion__headline ids-accordion__headline-1 ids-accordion__headline--m"
        >
          <h2>
            Rubrik 1
          </h2>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="ids-accordion__content"
        id=":r1:"
      >
        <div
          class="md:hidden"
        />
        <div
          class="hidden md:block"
        >
          <table
            class="ids-table"
          >
            <tbody>
              <tr>
                <td>
                  Test
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
